### ISSUE TITLE
dist_relevant_mse and val_dist_relevant_mse are nan during training 

### ISSUE THREAD

**Issue by anorak94** [mpicbg-csbd/stardist#55]
The dist_relevant_mse and val_dist_relevant_mse are nan during training when i try the same with a smaller training set of 10 images and validation set of 2 images this problem is not there and i manage to overfit my data, which i reckon is an indication that there is no problem with my images. so is this error because my training set is too large. I use 350 images for training and 50 for validation with no augmentation

**Comment by uschmidt83** [comment]
> The dist_relevant_mse and val_dist_relevant_mse are nan during training **when i try the same**

Did you intend to make a new issue or do you want to continue the discussion from #54?

**Comment by anorak94** [comment]
no this is a new issue right now i am training on 3 channel images and 1 channel masks although setting the envrionment variables using OMP_NUM_THREADS doesnt work for me

**Comment by anorak94** [comment]
i checked my output after training for 400 epochs and the network doesnt learn anything my output is completely blank

**Comment by maweigert** [comment]
> OMP_NUM_THREADS doesnt work for me

What does that mean? If there is still an issue with that, please add some info to #54 

> i checked my output after training for 400 epochs and the network doesnt learn anything my output is completely blank

Can you provide more information, e.g. share the training notebook (with its outputs)? It's impossible to give you an answer without knowing at least some details .

**Comment by anorak94** [comment]
https://we.tl/t-DRcTlZg4gQ

this is the link to the notebook

**Comment by anorak94** [comment]
here is the data im using 
https://we.tl/t-dF3y7ZnhOS

**Comment by maweigert** [comment]
Ok, thanks. 

1) A lot of your images (> 30%) have no annotations at all, so try to remove some of those...
2) add the following to your config

```python
conf = Config2D(
...
    train_foreground_only= 1,
    train_background_reg= 0 
...
```
This will disable the penalisation of background distances (which in your case swamps the loss)

**Comment by anorak94** [comment]
I have around 113 no annotations images and 250 images with annotations when i keep even 13 images without annotations i get nan in my training error when i remove all images with no annotations the problem with nans goes away. These are my metrics

`DatasetMatching(criterion='iou', thresh=0.4, fp=255, tp=902, fn=439, precision=0.7796024200518582, recall=0.6726323639075317, accuracy=0.5651629072681704, f1=0.722177742193755, n_true=1341, n_pred=1157, mean_true_score=0.46753341629243267, by_image=False)`

1) Will it help if i get more training images or train for longer durations. 
2) Shouldnt the network be able to distinguish patches with no annotations from patches with annotations. It seems like a shortcoming of the present method that its not able to do this (If my code is correct)
3) Are there other parameters that i can tune to increase my accuracy

**Comment by maweigert** [comment]
Hi, 

> I have around 113 no annotations images and 250 images with annotations when i keep even 13 images without annotations i get nan in my training error when i remove all images with no annotations the problem with nans goes away.

Your problem is that lot of your input images contain almost no image *signal*, but are of background or empty crops with just a few pixels having a high value. The percentile normalisation of these images then leads to these pixels having extremely large values (after normalisation), thus some parts of the losses become infinity.  

To get rid of these problems, consider using only images that contain some meaningful structures of your sample.

**Comment by anorak94** [comment]
hey thanks for all your help. the stardist approach is really cool. My metrics above are after removing all the zero annotation images. Is it possible to get better results by tweaking some of the config parameters in the stardist model. If not since stardist is based on. a Unet architecture would using different variants of the unet result in a significant improvement in accuracy. Because right now im not satisfied with the accuracy im getting

**Comment by maweigert** [comment]
> * Is it possible to get better results by tweaking some of the config parameters in the stardist
>   model.

I think the last config is fine. Is there a reason why you feed in 3 channel images? If there is a nuclear marker among them, it probably makes sense only to use that channel.   

> * If not since stardist is based on. a Unet architecture would using different variants of the unet
>   result in a significant improvement in accuracy.

Don't think that is an issue here. 

> * Because right now im not satisfied with the accuracy im getting or should i try different neural
>   networks such as mask-rcnn etc.

Sure, you could try Mask-RCNN (would be interesting to see the comparison).  

> * Also if i train for longer duration with more training data will that help

Improve the data itself (e.g. using only the nuclear channel) should have a bigger impact than training duration.
